modules/metrics/statistics.go

```diff
@@ -10,29 +10,72 @@ import (
 	"time"
 
 	"code.gitea.io/gitea/models"
-	"code.gitea.io/gitea/modules/cache"
-	"code.gitea.io/gitea/modules/setting"
 )
 
-var statisticsLock sync.Mutex
+var (
+	statisticsLock        sync.Mutex
+	statisticsMap         = map[string]*models.Statistic{}
+	statisticsWorkingChan = map[string]chan struct{}{}
+)
+
+func GetStatistic(estimate bool, statisticsTTL time.Duration, metrics bool) <-chan *models.Statistic {
```

Comment:
Why not use the common cache package?

Reply:
One of the problems of the common cache in `modules/cache` is that it cannot guarantee that a response will be there after it is put there. So in line 73 below we'd still need some local cache result. 

Now, could we do something clever with registering with the worker through a channel? Possibly, but I think this might be difficult to get right and small local cache is likely to be the least inefficient.

I can add pushing and querying of the cache service in addition if you think it's worth doing?

Reply:
This would be the way I'd add the common cache infrastructure:

```patch
diff --git a/modules/metrics/statistics.go b/modules/metrics/statistics.go
index 6d3695648..aa0dcbf94 100644
--- a/modules/metrics/statistics.go
+++ b/modules/metrics/statistics.go
@@ -12,7 +12,9 @@ import (
 
 	"code.gitea.io/gitea/models"
 	"code.gitea.io/gitea/models/db"
+	"code.gitea.io/gitea/modules/cache"
 	"code.gitea.io/gitea/modules/process"
+	"code.gitea.io/gitea/modules/setting"
 )
 
 var (
@@ -23,9 +25,19 @@ var (
 
 func GetStatistic(estimate bool, statisticsTTL time.Duration, metrics bool) <-chan *models.Statistic {
 	cacheKey := "models/statistic.Statistic." + strconv.FormatBool(estimate) + strconv.FormatBool(metrics)
+	ourChan := make(chan *models.Statistic, 1)
+
+	// First we should try the common cache
+	c := cache.GetCache()
+	if c != nil {
+		if stats, ok := c.Get(cacheKey).(*models.Statistic); ok && stats.Time.Add(statisticsTTL).After(time.Now()) {
+			ourChan <- stats
+			close(ourChan)
+			return ourChan
+		}
+	}
 
 	statisticsLock.Lock() // CAREFUL: no defer!
-	ourChan := make(chan *models.Statistic, 1)
 
 	// Check for a cached statistic
 	if statisticsTTL > 0 {
@@ -65,6 +77,11 @@ func GetStatistic(estimate bool, statisticsTTL time.Duration, metrics bool) <-ch
 			default:
 			}
 
+			// cache the result with the common cache
+			if c != nil && statsPtr != nil {
+				_ = c.Put(cacheKey, statsPtr, setting.DurationToCacheTTL(statisticsTTL))
+			}
+
 			// cache the result, remove this worker and inform anyone waiting we are done
 			statisticsLock.Lock() // Lock within goroutine
 			if statsPtr != nil {

```